Normative: Update calendar data consistency validation

gibson042 opened the pull request:
Fixes #2497
Fixes #2465

"era" and "eraYear" are still checked by PrepareTemporalFields in alphabetical order along with other fields when the calendar specifies their relevance, but ensuring both-or-neither and consistency between the pair and "year" is no longer part of the operation (although it remains part of CalendarResolveFields along with similar checks for "month" and "monthCode").

A followup should move Calendar-related operations from ECMA-402 to ECMA-262 after this and #2474... I initially included that here as 13e2c7328e473890c730d4b2735df18cc3e613f8, but backed it out to avoid potential merge conflicts.

ptomato commented on spec/intl.html:
```diff
@@ -1736,6 +1753,7 @@ <h1>Temporal.Calendar.prototype.yearMonthFromFields ( _fields_ [ , _options_ ] )
             1. If _calendar_.[[Identifier]] is *"iso8601"*, then
               1. Let _result_ be ? ISOYearMonthFromFields(_fields_, _overflow_).
             1. Else,
+              1. Perform ? CalendarResolveFields(_calendar_.[[Identifier]], _fields_, ~year-month~).
```
I was hoping that we didn't need to have different validation steps between dates and year-months, as we do for month-days, and I hadn't yet found any counterexample that would require it. Do you know of one, or is this split just in general?

gibson042 replied:
No, in practical implementations I expect CalendarResolveFields to be tightly integrated with CalendarDateToISO/CalendarMonthDayToISOReferenceDate and use identical logic for deciding whether to accept (and translate to ISO 8601) or reject a full date fields bag vs. a year–month fields bag. But this spec organization makes more sense for demonstrating parallels between the required ISO 8601 calendar and its dedicated operations vs. optional calendars with analogous but less-specified operations.

ptomato commented on spec/calendar.html:
```diff
@@ -711,29 +711,35 @@ <h1>
       </emu-alg>
     </emu-clause>
 
-    <emu-clause id="sec-temporal-resolveisomonth" aoid="ResolveISOMonth">
-      <h1>ResolveISOMonth ( _fields_ )</h1>
```
It would be good to perform this rename separately, for consistency with the other AOs named "ISO...", in the editorial PR, even if we decide not to fix #2497 in this way.

ptomato replied:
I've included this rename in https://github.com/tc39/proposal-temporal/pull/2577, rebasing once that is merged should help reduce the diff in this PR.

ptomato replied:
Rebased.

ptomato commented on spec/abstractops.html:
```diff
@@ -1660,6 +1660,7 @@ <h1>
         _fields_: an Object,
         _fieldNames_: a List of property names,
         _requiredFields_: ~partial~ or a List of property names,
+        optional _extraFieldDescriptors_: a List of Calendar Field Descriptor Records
```
#2472 also fixes #2465, and I hope to have that one merged before this normative change comes up for presentation in plenary. Is this fix incompatible with that PR or can it be rebased on top of that PR and/or folded into another commit in this PR?

gibson042 replied:
Handled by rebase; it's still a "Normative: Remove the ECMA-402 PrepareTemporalFields override" commit in this PR.

ptomato reviewed (commented):
The editorial changes seem good in any case, could we get them in a separate PR?

In particular, "Editorial: Consolidate standard-field Calendar method logic into ECMA-262" is #1418, which I was planning on working on after getting all the normative issues finished. I realize it's later reverted in this PR, but it'd be great to put that work towards addressing that issue in a separate PR!

This looks to be less invasive than I thought it would be, so I'm more positive about it than I expected. It'd be helpful to me, if I could revisit it again once it's rebased.

gibson042 commented:
I rebased on main and extracted the initial sequence of editorial commits into #2506, upon which this PR now sits ([diff](https://github.com/gibson042/proposal-temporal/compare/2023-02-calendar-op-consistency...gibson042:proposal-temporal:gh-2466-calendar-field-validation)). Ephemeral consolidation of Calendar method logic into ECMA-262 in 13e2c7328e473890c730d4b2735df18cc3e613f8 was lost in the process due to merge conflicts, so I'm explicitly mentioning it here.

Aditi-1400 commented on spec/intl.html:
```diff
@@ -2210,10 +2213,12 @@ <h1>Temporal.Calendar.prototype.fields ( _fields_ )</h1>
-              1. Let _result_ be _fieldNames_.
-            1. Else,
-              1. Let _result_ be CalendarDateFields(_calendar_.[[Identifier]], _fieldNames_).
+            1. Let _result_ be _fieldNames_.
+            1. If _calendar_.[[Identifier]] is not *"iso8601"*, then
+              1. NOTE: Every built-in calendar preserves all input field names in output.
+              1. Let _extraFieldDescriptors_ be CalendarFieldDescriptors(_calendar_.[[Identifier]], ~date~).
+              1. For each Calendar Field Descriptor Record _desc_ of _extraFieldDescriptors_, do
+                1. Append _desc_.[[Property]] to _result_.
```
@gibson042, me and @ptomato were discussing this change for polyfill changes and we believe, these steps might be slightly incorrect. 
Previously, `calendar.fields(['monthCode', 'day'])` would have returned `['monthCode', 'day']` but with this change it would return `['monthCode', 'day', 'era', 'eraYear']`

We believe there should be a check for determining the _type_ from the input to the fields! 
What do you think?

gibson042 replied:
I agree that this PR should not affect the output `builtinNonISOCalendar.fields(['monthCode', 'day'])`, but wasn't happy with the specific changes so I have moved derivation logic back into `CalendarFieldDescriptors` for use only by `Temporal.Calendar.prototype.fields`.

ptomato replied:
This seems like a good solution; keep `~date~`, `~year-month~`, and `~month-day~` for `dateFromFields()`, `yearMonthFromFields()`, and `monthDayFromFields()` respectively; and pass the array directly for `fields()`.

@Aditi-1400 Even though it's a while ago, you might still have more insight into this than I do; want to take a look?

justingrant reviewed (commented):
I think this looks good, but I didn't spend enough time with it to be 100% confident so leaving for others to approve. A few minor comments below.

ptomato commented:
> What's the plan with this PR? It's been open for a very long time. Is it still relevant?

It's waiting on resolution of https://github.com/tc39/proposal-temporal/pull/2500#discussion_r1184400810

justingrant commented:
> > What's the plan with this PR? It's been open for a very long time. Is it still relevant?
> 
> 
> 
> It's waiting on resolution of https://github.com/tc39/proposal-temporal/pull/2500#discussion_r1184400810

@gibson042, possible to respond to that comment so we can unblock this? Thx!

gibson042 commented:
Updated and rebased onto main.

justingrant commented on spec/calendar.html:
```diff
@@ -834,22 +834,22 @@ <h1>
         <dt>description</dt>
         <dd>
-          It validates that the *"month"* and *"monthCode"* properties of the given _fields_ are not inconsistent and merges them into an integer month which is returned, or throws an exception if both are *undefined*.
+          It validates that the *"month"* and *"monthCode"* properties of the given _fields_ are not inconsistent and merges them into a *"month"* property with a positive integral Number value in the inclusive interval from 1 to 12, or throws an exception if both are *undefined*.
```
To me, "merges" implies that `monthCode` will be removed from the object, or at least is somewhat ambiguous about what will happen with `monthCode` after the AO returns. Maybe something like the following would be clearer?

```suggestion
          It ensures that the *"month"* property of _fields_, if *undefined*, is set to the integral Number value in the inclusive interval from 1 to 12 that corresponds to the *"monthCode"* property of  _fields_.
          If both properties are *undefined* or if the properties are inconsistent, or if either property is invalid, then a *RangeError* will be thrown.
```

ptomato commented on polyfill/lib/calendar.mjs:
```diff
@@ -83,7 +85,7 @@ export class Calendar {
     if (!ES.IsTemporalCalendar(this)) throw new TypeError('invalid receiver');
     const fieldsArray = [];
-    const allowed = new Set([
+    const allowed = new OriginalSet([
       'year',
       'month',
       'monthCode',
```
I think this is where the only rebase conflict is; you'll have to remove the time units from this list.

ptomato reviewed (approved):
Works for me, I'm fine to merge it along with https://github.com/tc39/test262/pull/3835 when other reviewers are satisfied.

gibson042 commented:
Rebased.

Aditi-1400 reviewed (approved):
Looks good to me! :)
